### PR TITLE
[RANS] QSVMS Expansion Step 1 - Add Constitutive Law Support

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This is the first step towards expanding RANS to support QSVMS (And all other stabilization methods if requred). This introduces templated `RansConstitutiveLawExtension` which can be used to extend easily `FluidConstitutiveLaw` just by providing the type of it as a template argument.

**Changelog**
- Adds `RansConstitutiveLawExtension`
- Adds `RansNewtonian2DLaw` as used example
- Adds `RansNewtonian3DLaw` as used example

### CODE REVIEW

**Review by rubenzorrilla — COMMENTED** [review]
Long story short, we need this to customize the effective viscosity calculation but keeping the rest of the constitutive law ¿right?. Honestly, I'm not very positive about introducing a template argument in the constitutive law after the experience of the mechanical ones. 

In this regard, I note that we are facing a very similar issue with the artificial magnitudes (@pau-mar @RiccardoRossi). So far we decided to duplicate the constitutive law, which is without discussion a temporary solution.

I want to confirm that this is the problematic and to better discuss it before merging.

**Comment by sunethwarna** [comment]
> Long story short, we need this to customize the effective viscosity calculation but keeping the rest of the constitutive law ¿right?. 

This is exactly what I want. I am open to suggestions. This was the only way i found which is easier to look at, extend and manage `FluidConstitutiveLaws`. Constitutive laws are so complicated, and has lots of elements containers assigned to them, and I don't know how to manage them if we make duplicates everywhere :/

> I want to confirm that this is the problematic and to better discuss it before merging.

I agree. Even ConstitutiveLaw::Parameters are also kind of complicated, and I don't totally get the signature of the `ConstitutiveLaw::CalculateValue` also. I would like to hear all the opinions :)

**Comment by RiccardoRossi** [comment]
you can get documentation of the CL in the wiki. In my opinion using the CL is a must...

**Comment by sunethwarna** [comment]
@RiccardoRossi thanks :).... A small question, I'm just curious, why does `ConstitutiveLaw::CalculateValue` take output value as a reference and as well as have the return type of the same? (In the followgin takes `double&` and returns a double as well.

https://github.com/KratosMultiphysics/Kratos/blob/261a51e19b78fd547466256e557e88627e59e7f5/kratos/includes/constitutive_law.h#L745

**Review by rubenzorrilla — COMMENTED** [review]
But why don't deriving them from the standard newtonian laws? At the end we just need to `override` the effective viscosity calculation method and enhance the check one to be sure that the `TURBULENT_VISCOSITY` is in the nodal data.

**Comment by sunethwarna** [comment]
Then its gonna be same as the previous approach, (The template paramter was made the base class) I thought that is what you wanted to avoid, have dependencies which makes developement harder.

Previously what I had was something like this
```c++
template<class TBaseClassType>
class RansConstitutiveLaw : public TBaseClassType
{
protected:
      // overriding the GetEffectiveViscosityMethod
};

template class RansConstitutiveLaw<Newtonian2DLaw>;
```

And proposed deriving also will have the same (apart from the template argument)
```c++
class RansConstitutive2DLaw : public Newtonian2DLaw
{
protected:
      // overriding the GetEffectiveViscosityMethod
};
```

I dont see much difference in two methods
1. 1st method will only require `RansConstitutiveLaw` class (only two files)
2. 2nd method will require two classes 'RansConstitutive2DLaw` and `RansConstitutiveLaw` (needs 4 files)

Honestly I don't see a difference in two aproaches. What do you think? (Thats why recreated the whole laws from scratch so it can be decoupled for furture developements and expansions - current state of this PR)

**Comment by rubenzorrilla** [comment]
The point is that I can't foresee so much room for further developments and expansions for a Newtonian law. The idea of method 2 is precisely that, avoid the template argument. Note that as it is right now it requires 4 files but only one method `override`, which can be eventually implemented in an external utility. Also note that the constitutive laws in the `FluidDynamicsApplication` could be eventually templated with the dimension. This is not done yet because of legacy reasons (some of the non-Newtonian constitutive models have no 2D version yet). By doing so we will end up with a two files implementation as option one.

**Comment by sunethwarna** [comment]
@rubenzorrilla Done. Could you have a look?

File: applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+int RansNewtonian2DLaw::Check(
+    const Properties& rMaterialProperties,
+    const GeometryType& rElementGeometry,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    // Check viscosity value
+    KRATOS_ERROR_IF(rMaterialProperties[DYNAMIC_VISCOSITY] <= 0.0)
+        << "Incorrect or missing DYNAMIC_VISCOSITY provided in process info "
```
Comment (rubenzorrilla) [inline applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.cpp]:
```suggestion
        << "Incorrect or missing DYNAMIC_VISCOSITY provided in material properties "
```

File: applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+           "for RansNewtonian2DLaw: "
+        << rMaterialProperties[DYNAMIC_VISCOSITY] << std::endl;
+
+    KRATOS_ERROR_IF(rMaterialProperties[DENSITY] <= 0.0)
+        << "Incorrect or missing DENSITY provided in process info "
```
Comment (rubenzorrilla) [inline applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.cpp]:
```suggestion
        << "Incorrect or missing DENSITY provided in material properties "
```
(I've just realised that this is also wrong in the Newtonian law)

File: applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.cpp (hunk 1 above)
Comment (sunethwarna) [reply applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.cpp]: :/ copy paste mistake... hehe... thanks :)

File: applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * Copy constructor.
+     */
+    RansNewtonian2DLaw(const RansNewtonian2DLaw& rOther);
+
+    /**
+     * Destructor.
+     */
+    ~RansNewtonian2DLaw() override;
```
Comment (rubenzorrilla) [inline applications/RANSApplication/custom_constitutive/rans_newtonian_2d_law.h]:
```suggestion
    ~RansNewtonian2DLaw();
```

File: applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+int RansNewtonian3DLaw::Check(
+    const Properties& rMaterialProperties,
+    const GeometryType& rElementGeometry,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    // Check viscosity value
+    KRATOS_ERROR_IF(rMaterialProperties[DYNAMIC_VISCOSITY] <= 0.0)
+        << "Incorrect or missing DYNAMIC_VISCOSITY provided in process info "
```
Comment (rubenzorrilla) [inline applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.cpp]:
```suggestion
        << "Incorrect or missing DYNAMIC_VISCOSITY provided in material properties "
```

File: applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+           "for RansNewtonian3DLaw: "
+        << rMaterialProperties[DYNAMIC_VISCOSITY] << std::endl;
+
+    KRATOS_ERROR_IF(rMaterialProperties[DENSITY] <= 0.0)
+        << "Incorrect or missing DENSITY provided in process info "
```
Comment (rubenzorrilla) [inline applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.cpp]:
```suggestion
        << "Incorrect or missing DENSITY provided in material properties "
```

File: applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * Copy constructor.
+     */
+    RansNewtonian3DLaw(const RansNewtonian3DLaw& rOther);
+
+    /**
+     * Destructor.
+     */
+    ~RansNewtonian3DLaw() override;
```
Comment (rubenzorrilla) [inline applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.h]:
```suggestion
    ~RansNewtonian3DLaw();
```

**Review by rubenzorrilla — COMMENTED** [review]
Just a few minor comments.

File: applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.h (hunk 6 above)
Comment (sunethwarna) [reply applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.h]: Don't we need override here? (we have virtual method)

Comment (rubenzorrilla) [reply applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.h]: This is not a standard method, is the destructor. https://github.com/isocpp/CppCoreGuidelines/issues/721

Comment (sunethwarna) [reply applications/RANSApplication/custom_constitutive/rans_newtonian_3d_law.h]: done :)

**Comment by sunethwarna** [comment]
Hi @rubenzorrilla , I addressed all of your comments. Could you have a look?

**Review by rubenzorrilla — APPROVED** [review]

**Comment by sunethwarna** [comment]
thanks @rubenzorrilla :)

**Pull request merged**